src/db_iterator.rs

```diff
@@ -67,14 +67,21 @@ use std::slice;
 /// ```
 pub struct DBRawIterator<'a> {
     inner: *mut ffi::rocksdb_iterator_t,
+
+    /// When iterate_upper_bound is set, the inner C iterator keeps a pointer to the upper bound
+    /// inside `_readopts`. Storing this makes sure the upper bound is always alive when the
+    /// iterator is being used.
+    _readopts: ReadOptions,
```

Comment:
I don't understand why `ReadOptions` needs to outlive the iterator, it seems to me that only `upper_bound` needs that.
You can see e.g. that this test passes
```rust
#[test]
fn iterator_test_upper_bound() {
    let path = "_rust_rocksdb_iteratortest_upper_bound";
    {
        let mut opts = Options::default();
        opts.create_if_missing(true);
        opts.create_missing_column_families(true);
 
        let db = DB::open_cf(&opts, path, &["cf1"]).unwrap();
        let cf1 = db.cf_handle("cf1").unwrap();
        let key1 = b"0";
        let key2 = b"ab";
        let key3 = b"abc";
        let key4 = b"abcd";
 
        db.put_cf(&cf1, key1, key1).unwrap();
        db.put_cf(&cf1, key2, key2).unwrap();
        db.put_cf(&cf1, key3, key3).unwrap();
        db.put_cf(&cf1, key3, key4).unwrap();
 
        let mut readopts = ReadOptions::default();
        unsafe { readopts.set_iterate_upper_bound(b"1"); }
 
        let iter = db.iterator_cf_opt(cf1, &readopts, IteratorMode::From(key1, Direction::Forward));
        drop(readopts);
        let expected: Vec<_> = vec![(key1, key1)]
            .into_iter()
            .map(|(k, v)| (k.to_vec().into_boxed_slice(), v.to_vec().into_boxed_slice()))
            .collect();
        assert_eq!(expected, iter.collect::<Vec<_>>());
    }
    let opts = Options::default();
    DB::destroy(&opts, path).unwrap();
}
```

Reply:
It's true that only `upper_bound` needs that and this test passes. However, the goal of this PR is to make `set_iterate_upper_bound` safe, so users can use this API without writing unsafe code, and reduce the possibility of user messing up by calling this API incorrectly. And this test above doesn't show how we can achieve that goal without making any changes to the APIs.

Let's say we do not make changes to the existing APIs, so user would still pass a reference to functions like `iterator_opt`. Can you elaborate how the user code would look like in this approach?

Reply:
Fair enough, I haven't really thought this through. One alternative would be to tie the lifetime of `&ReadOptions` to the iterator (`&'b ReadOptions` or `&'a ReadOptions`), but I'm not sure it's better than passing it by value. The problem with passing it by value though is that a user can't reuse the same `ReadOptions` as it's not clonable.

Or maybe it's possible to do some trickery with `Pin<Box<[u8]>>`?

Reply:
Another idea is to store the upper_bound as `Option<Arc<[u8]>>` in both `ReadOptions` and the iterator. I think this could work.

Reply:
Yeah, I thought about doing the lifetime stuff, but I kind of feel that it's not worth the effort. Adding lifetime restriction is not going to be a backward compatible change anyway, since the users might have the following code:

```
struct MyDB {
    inner: rocksdb::DB,
}

impl MyDB {
    fn get_iter(&self) -> DBIterator {
        let readopts = ...;
        DB::iterator_opt(&readopts)
    }
}
```

Also note that the snapshot APIs take read options by value: https://github.com/rust-rocksdb/rust-rocksdb/blob/0e84c3246844d94efcbab0f94915654c3828072a/src/snapshot.rs#L55 So we could make some consistency improvements :) I honestly think passing by value is gonna be a small change on the user side.

I'll give the `Arc` idea a try and see how the code looks like.